Feature/orchestrator balance check

free2fqn opened the pull request:
Запилил вьюшку для баланса с ссылочкой на его пополнение.

free2fqn committed:
added fields sms.url.replenish-balance, sms.url.get-balance, sms.api.id to application.yml

free2fqn committed:
created AdminController

free2fqn committed:
added balance-check.html view. Created AdminController. Reworked NotificatorFeign. Deleted SampleController

nekromant322 reviewed (changes requested):
api для доступа к балансу

free2fqn committed:
deleted AdminController: all endpoints now are in PageController. Created utility endpoint getBalance and view-returning balancePage. Returned balance-check.js. Reworked balance-check.html, removed Thymeleaf

free2fqn commented:
@nekromant322 пофиксил, но не уверен, что хорошо складывать все эндпоинты в PageController. Возможно нужен отдельный еще один rest-controller?

nekromant322 commented on platform_orchestrator/src/main/java/com/override/controller/PageController.java:
```diff
@@ -16,4 +28,15 @@ public String coursesPage() {
     }
 
+    @ResponseBody
+    @GetMapping("/admin/rest/balance")
```
Сюда складывать не надо, этот контроллер только чтоб возвращать пустые странички до появления отдельного "сервера" с фронтом, вынести в отдельный контроллер со звучным названием и убери зависимость на notificationFeign

nekromant322 commented on platform_orchestrator/src/main/resources/application.yml:
```diff
@@ -10,6 +10,13 @@ spring:
   application:
     name: orchestrator
 
+sms:
+  url:
+    replenish-balance: https://sms.ru/pay.php
+    get-balance: https://sms.ru/my/balance?api_id=
+  api:
+    id: ${SMS_API_ID}
```
зачем здесь? оставь только ссылку на пополнение

nekromant322 commented on platform_orchestrator/src/main/java/com/override/controller/PageController.java:
```diff
@@ -1,11 +1,23 @@
 import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
 import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestBody;
```
ненужные импорты

nekromant322 commented on platform_orchestrator/src/main/java/com/override/controller/PageController.java:
```diff
@@ -16,4 +28,15 @@ public String coursesPage() {
+        return notificatorFeign.getBalance();
+    }
+
+    @GetMapping("admin/balancePage")
```
вынести в отдельный контроллер, не использовать префикс /admin, потом в конфиге секурити паттернами ендпоинты засекурятся

nekromant322 commented on platform_orchestrator/src/main/resources/templates/balance-check.html:
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/html">
+<head>
+    <!-- Обязательные метатеги -->
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+
+    <!-- Bootstrap CSS -->
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
+
+    <title>Balance</title>
+</head>
+<body>
+  <h3>Balance on sms.ru</h3>
+  <label for="balanceText">Balance(RUB): </label>
+  <input id="balanceText" disabled/>
+
+  <a href="https://sms.ru/pay.php" class="link-success">Replenish the balance</a>
+  <!-- Вариант 1: Bootstrap в связке с Popper -->
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.0/jquery.min.js"></script>
+  <script src="/js/balance-check.js"></script>
+</body>
+</html>
```
можно бы добавить отдельный хедер для страничек админа, где была бы навигация для остальных ютилити штук типо просмотра отчетов и еще всякого разного

nekromant322 replied:
пример можно посмотреть в папке fragments

nekromant322 commented:
как на счет обработать ситуацию аля "сервис нотификаций не отвечает" или "сервис отвечает, но баланс получить все равно не смог" (отвечает 500)

free2fqn committed:
Cleaned up application.yml.
Cleaned up PageController, reworked "/balancePage" endpoint, "/balance" endpoint and NotificatorFeign moved to NotificationController.
Created NotificationController.
Now "href" attribute links to sms.url.replenish-balance field in application.yml via endpoint "/balance" in NotificationController.
Added second ajax in balance-check.js to get value for "href" for replenish-balance-link.

free2fqn committed:
Added navbar in balance-check.html.

free2fqn committed:
Added logger in NotificationController. If notification-service is unavailable log will write it, and getBalance will return "Service unavailable".

free2fqn commented:
> как на счет обработать ситуацию аля "сервис нотификаций не отвечает" или "сервис отвечает, но баланс получить все равно не смог" (отвечает 500)

теперь логируется эксепшен, когда сервис нотификаций не отвечает

nekromant322 commented on platform_orchestrator/src/main/java/com/override/controller/NotificationController.java:
```diff
@@ -0,0 +1,32 @@
+
+    @GetMapping("/url/replenish")
+    public String getUrlToReplenishBalance() {
+        return urlToReplenishBalance;
```
как на счет объекдинить в дто и взвращать вместе с балансом? зачем лишние вызовы на бэк делать?

nekromant322 commented on platform_orchestrator/src/main/resources/templates/balance-check.html:
```diff
@@ -9,6 +9,23 @@
     <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3" crossorigin="anonymous">
 
     <title>Balance</title>
+
+    <nav class="navbar navbar-dark bg-dark">
+        <!-- Navbar content -->
+        <li class="nav-item">
+            <a class="nav-link" href="/courses">Курсы</a>
+        </li>
+        <li class="nav-item">
+            <a class="nav-link" href="/">Index</a>
+        </li>
+        <li class="nav-item">
+            <a class="nav-link" href="#">Link</a>
+        </li>
+        <li class="nav-item">
+            <a class="nav-link" href="#">Link</a>
+        </li>
+    </nav>
```
сделать фрагментом, чтоб подключать на другие страницы админа удобно было, см пример в папке fragments

nekromant322 commented on platform_orchestrator/src/main/java/com/override/controller/NotificationController.java:
```diff
@@ -22,7 +25,14 @@ public NotificationController(NotificatorFeign notificatorFeign) {
+            balance = "Service unavailable";
+            log.error("Caught FeignException: " + e.getMessage());
+        }
+        return balance;
```
не уверен, что что контроллер предназначен для этого  
еще я бы бросал все таки 500, тогда фронту будет легко понять что все наебнулось  (с адекватным сообщением об ошибке, разумеется)
А собственно сама обработка пусть будет на фронте (получили 500 - нарисовали красную надпись)

nekromant322 committed:
error handling advice

nekromant322 committed:
error handling advice

free2fqn committed:
Merge branch 'feature/error-handling' of https://github.com/nekromant322/platform into feature/orchestrator-balance-check

free2fqn committed:
Added BalanceResponseFromNotificationControllerDTO. BalanceResponseDTO renamed in BalanceResponseFromSmsRuFeignClientDTO.
Type of all "balance" fields and methods changed to double in: BalanceResponseFromSmsRuFeignClientDTO, CodeCallController, CodeCallService, CodeCallServiceImpl, NotificationController.
Removed url field in CodeCallServiceImpl.

free2fqn committed:
Added @AllArgsConstructor in BalanceResponseFromNotificationControllerDTO.
Now endpoint "/balance" in NotificationController returns BalanceResponseFromNotificationControllerDTO.
Reworked initBalance() in balance-check.js for single AJAX.

free2fqn committed:
Added admin-navbar.html

nekromant322 commented on platform_notification/src/main/java/com/override/feign/SmsRuFeign.java:
```diff
@@ -12,5 +12,5 @@
 
     @GetMapping("/my/balance?api_id={apiID}&json=1")
-    BalanceResponseDTO getBalance(@PathVariable String apiID);
+    BalanceResponseFromSmsRuFeignClientDTO getBalance(@PathVariable String apiID);
```
-> `SmsRuBalanceResponseDTO`

nekromant322 commented on platform_orchestrator/src/main/resources/templates/fragments/admin-navbar.html:
```diff
@@ -0,0 +1,24 @@
+<body>
+<header th:fragment="header">
+    <nav class="navbar navbar-dark bg-dark">
+        <!-- Navbar content -->
+        <li class="nav-item">
+            <a class="nav-link" href="/courses">Курсы</a>
+        </li>
+        <li class="nav-item">
```
а где линк на проверку баланса в хедере админа?

free2fqn committed:
Added link to balance page in admin-navbar.html
Clean up balance-check.js
Rename BalanceResponseFromSmsRuFeignClientDTO to SmsRuBalanceResponseDTO
